### DISCUSSION TITLE
Adjusting the applicability of composite rules

### DISCUSSION THREAD

**Discussion by WilcoFiers** [auto-wcag/auto-wcag#2214] [General]
This topic has been discussed a few times now. The problem we were looking at is this one. WCAG's target size requirement is fairly complex. It has a the a number of clauses, and a number of exceptions. In order to avoid having to put all of this complexity into a single rule, we'd like to break different parts of it up and put those into atomic rules, then have one composite rule for each success criterion to put it together.

Under the current ACT Rules Format, this could look like this:

## Option 1: current proposal

```md
- Target has minimal size or spacing (composite) -- This rule passes if one of the following passes
    - Target has minimal size (atomic)
    - Target has sufficient space around it (atomic)
    - Target is in a text block (atomic)
    - Target size is controlled by the user agent (atomic)
    - Target size is essential or has an alternative (atomic)
```

The applicability of these three rules will then be something like this:

```md
Each target of a widget in the page.
```

**Drawback** This creates a problem where small targets will end up passing the rule if they are in a block of text, controlled by the user agent, or the size is essential. The language of the success criterion seems to suggest this should be inapplicable. While in practice this doesn't change whether a page conforms to WCAG or not, passing things that seem like they should be inapplicable has proven to be confusing to people in the past. This is something we'd prefer to avoid.

Putting these exceptions into their own atomic rules can be especially confusing, since these atomic rules failing things that are in no way an accessibility problem. The inline text block atomic rule fails controls that are not in line. The user agent size rule fails controls who's size is not controlled by the user agent, etc.

To address this, we came up with a few options:

## Option 2: Definitions

The most obvious solution is to take the things that should be inapplicable, and rather than put them in a separate rule we put those ideas into a definition which is used in the applicability of both the atomic and the composite rules. The rule structure would then be:

```md
- Target has minimal size or spacing (composite) -- This rule passes if one of the following passes
    - Target has minimal size (atomic)
    - Target has sufficient space around it (atomic)
```

And the applicability of these three rules will then be something like this:

```md
Each target of a widget in the page, except if one of the following is true:
  - The [target is in a block of text][]; or
  - the [size of the target is set by the user agent][]; or
  - an [equivalent exists in the page][]; or;
  - The target size is [essential][].
```

**Drawbacks**: The the atomic rules this results in cannot be reused between Target Size (Minimum) and Target Size (Enhanced). Since Target Size (Enhanced) does not allow for spacing, it wouldn't need to leaves only one requirement, which means no composite rule is necessary. Whether Target Size (minimum) need atomic rules is questionable then too.

## Option 3: Adjusting applicability in composite rules

Another thing that has been proposed is to allow the composite rule to limit the applicability. This is similar to option 2, except that the applicability of the composite rule would be different from that of the atomic rules:

Atomic rule applicability:

```md
Each target of a widget in the page.
```

Composite rule applicability: 

```md
Each target of a widget in the page, except if one of the following is true:
  - The [target is in a block of text][]; or
  - the [size of the target is set by the user agent][]; or
  - an [equivalent exists in the page][]; or;
  - The target size is [essential][].
```

**Drawbacks**: The big problem with this is that it is a significant change to the ACT Rules Format. Currently the only thing that can be used to determine the outcome in a composite rule is the outcome of the composite rules. If we update the rules format to allow the applicability of the composite rule to change, we need to add input aspect in the composite rule. This fundamentally changes what composite rules do, and as a result may make rules written in the ACT Rules Format more difficult to implement in tools using different architectures.

## Option 4: Atomic outcomes in the applicability

This approach is has all the atomic rules from option 1, except that it uses some of the outcomes in the applicability, rather than using all in the expectation:

Composite rule applicability:

```md
Each target of a widget in the page, except one of the following atomic rules passed:
  - Target is in a text block (atomic)
  - Target size is controlled by the user agent (atomic)
  - Target size is essential or has an alternative (atomic)
```

Composite rule expectation:

```md
At least one of the following rule passes:
  - Target has minimal size (atomic)
  - Target has sufficient space around it (atomic)
```

**Drawbacks**: This is currently not allowed by the ACT Rules Format, although since this won't require input aspects to be added to the composite rule, it is a smaller change to the rules format than option 3, and will likely be less impactful on implementors. And while this partially addresses the issue that the overall outcome will better match the expected outcome, this solution still requires atomic rules that fail things that are not accessibility problems. (I.e failing controls that are not inline text, who's size is not controlled by the user agent, etc.)

## Task force conclusions

Looking at all these solutions the ACT Task Force seemed to lean towards option 2. Option 1 and 4 have atomic rules that seem "off" for failing things that aren't accessibility problems, and option 3 doesn't seem like it's a enough improvement over option 2 to justify the changing the rules format in a way that potentially makes implementation more difficult.

I've opened this discussion for people to share alternative proposals, or if someone would like to make a case in favor of something other than option 2.

**Comment by Jym77** [comment]
Thread to discuss **Option 1: current proposal**.

**Reply by Jym77** [reply]
* I agree on the confusion in "user-agent controlled should be inapplicable, not passed". Especially since we've witnessed it a couple of times (including using dirty tricks ourselves to move the subjective parts of Applicability in the Expectation, see #2061). However, in that case 2.5.8 list "**Spacing**" exactly the same way as the other 4 exceptions. While I totally agree with the feeling that Spacing should be an expectation while "User-Agent controlled" should be an (in)applicability, I feel the normative language doesn't support it. This doesn't really change the bottom line of the confusion created by the current proposal…
* I am, however, less sure about the problem with "atomic rules that seem 'off' for failing things that aren't accessibility problems". This is pretty much what we have for all our atomic rules (since Composites are currently only "or"). E.g. [Video element visual-only content has transcript](https://www.w3.org/WAI/standards-guidelines/act/rules/ee13b5/proposed/) is failing videos without transcript even if they have, e.g., an audio track alternative. And in any case, if "size" and "spacing" are two atomic rule, "size" would fail stuff that are too small but spaced enough, which is also similar. 
    As far as I remember, this situation causes a bit of problems in making it harder for tools that don't follow the same split to claim an implementation of the atomic rules (and thus unfairly decrease their implementation score), which we solved by asking that the test cases in atomic rules also fail the SC their are meant for.

**Comment by Jym77** [comment]
Thread to discuss **Option 2: Definitions**.

**Reply by Jym77** [reply]
There are two problems with that, specifically for this SC:
* ~Some of these definitions are subjective (typically "essential size"). So they are currently not allowed in Applicability yet. This is somewhat minor since we do want to update that anyway.~ My bad, the format has already been updated 🙈 
* The "equivalent" condition actually recursively include the full SC 🙈 A too small widget can pass 2.5.8 by having an equivalent one that is spaced enough. This is not a huge problem per se (which just need two definitions of "an equivalent that is good", the same way #2167 has two "equivalent" atomic rules. But it does make things a bit messy and notably these definitions are ugly since they themselves need to include the other exceptions (a widget can pass by having an equivalent one that is UA-controlled).

**Comment by Jym77** [comment]
Thread to discuss **Option 3: Adjusting applicability in composite rules**.

**Reply by Jym77** [reply]
Yup, I don't like it either. Requesting Input Aspects in a composite rule is straying away from what they are meant to be: a combination of results from atomic rules…

**Comment by Jym77** [comment]
Thread to discuss **Option 4: Atomic outcomes in the applicability**.

**Reply by Jym77** [reply]
I actually like this option a lot, maybe by pushing it even further.

The core idea in the atomic/composite split is that a composite rule should not (have to) look at the actual content and can work by only looking at the results of the atomic rules. From a typing/function point of view, we have things like (this is very much influenced by Alfa's implementation of the ACT rules format):
```
atomic.applicability: Subject => Iterable<Target>
atomic.expectation: Target => "passed" | "failed"   // ignoring "can't tell" here
atomic.evaluate: Subject => { passed: Iterable<Target>; failed: Iterable<Target> }
```
Now, the way I understand the concept of composite rule is that it never really look at the `Subject`. It only calls the evaluation of each of its atomic rules and then combines the resulting sets in whichever clever way it does to get its own lists of passed and failed outcomes.
The Applicability of composite rules is (currently) enforced as being the union of all the Applicability of the atomic rules in the ACT rules format, and the format allows to skip it. I think all our composite rules currently have an explicit Applicability which is also a `Subject => Iterable<Target>` description. In a sense, this doesn't follow the spirit of the format as it does look into the Subject, and we could incorrectly write an Applicability different from the ones of the atomic rules (i.e. already do Option 3, it is not allowed by the format but if we forget to update something, it may slip in). (note: Alfa does indeed just build the union of Applicabilities).

---
Now, where I want to come is that I feel we get **closer** to the spirit of the format (at least how I see it) by not only (i) allowing Applicability of composite rules to be a logical combination of the outcomes of the atomic rules (this proposal); but actually (ii) **forcing** the Applicability of composite rules to be a logical combination of the outcomes of atomic rules (forbiding anything that looks into the subject).

So, for this Target Size rule, I'd go with an Applicability like:
> This rule applies to anything that fails all of (i) rule "user-agent controlled"; (ii) rule "essential size"; …

with expectation

> The outcome of at least one of (i) rule "large enough"; (ii) rule "spaced enough" is `passed`.

Then, it is obvious that the composite rule is not looking at the subject at all and is only making combination of outcomes of atomic rules. The composite rule doesn't even need to know what a widget is or what not, it is relying on the atomic rules to do so.

Existing composite rules could be updated the same way. E.g., [Video element visual content has accessible alternative](https://www.w3.org/WAI/standards-guidelines/act/rules/c5a4ea/proposed/) could switch its Applicability from
> This rule applies to every non-streaming `video` element that is visible, where the video contains audio.

(which makes it look like it does look into the subject and needs input aspects. And may break the format if we update one of the atomic rules to a different Applicability and forget to update this one)

to
> This rules applies to any element that is Applicable to any of:
> * Video Element Visual Content Has Audio Description
> * Video Element Visual Content Has Transcript
> * Video Element Content Is Media Alternative For Text

(which makes it obvious that we are indeed following the format and using the union of the Applicabilities of the atomic rules)

So, this is my preferred option, by taking it one step further and moving all composite rules a bit more into the "abstract" side of things.

It does still have the "problem" of atomic rules that fails stuff that are not accessibility problems, but as I mentioned in Option 1, this is already the case for all of our atomic rules. Currently, atomic rules are one way to pass a SC and failing them never guarantees the SC fails.

**Comment by Jym77** [comment]
Side note: while the discussion is focused here on the recent "Target Size" proposal, I feel that the core of it is broadly relevant. Notably:
* [Block of repeated content is collapsible](https://www.w3.org/WAI/standards-guidelines/act/rules/3e12e1/proposed/) (used in the Bypass Blocks rule) has a nasty "for each block of repeated content" hidden in its expectation. Which means that pages without repeated content are passing it (and the composite Bypass Block rule) while it feel that they should be Inapplicable for 2.4.1
* Nearly all media rules have a variant of [Video element content is media alternative for text](https://www.w3.org/WAI/standards-guidelines/act/rules/ab4d13/proposed/). Here also, I feel that videos that are media alternative for text should be Inapplicable for 1.2.* rather than Passed.

So, I think this is a fairly important discussion (not just local to a proposal). And we can also try the various options on existing and approved rules…
